adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -120,13 +116,548 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+        .map(f => (f._1, ((f._2.head._2._1, f._2.last._2._1), f._2.map(g => (g._1, g._2._2))))).toIterator
+    }
+
+    /**
+     * The purpose of this method is to co-partition two RDDs according to their ReferenceRegions.
+     * It is useful during ShuffleRegionJoins because it guarantees no extra shuffling to get correct
+     * and complete region joins. RDD is already sorted coming in, so we take precaution to keep it
+     * that way.
+     *
+     * The simplest way of sending data to the correct node is to just send everything that has a
+     * start position falls within the boundaries
+     *
+     * This is best used under the condition that (repeatedly) repartitioning is more expensive than
+     * calculating the proper location of the records of this.rdd. It requires a pass through the
+     * co-located RDD to get the correct partition(s) for each record. It will assign a record to
+     * multiple partitions if necessary.
+     *
+     * @param rddToCoPartitionWith the rdd that this.rdd will be copartitioned with
+     * @return returns the newly repartitioned rdd that is co-partitioned with rddToCoPartitionWith
+     */
+    def coPartitionByGenomicRegion[X, Y <: GenomicRDD[X, Y]](rddToCoPartitionWith: GenomicRDD[X, Y])(implicit tTag: ClassTag[T], xTag: ClassTag[X]): U = {
+      require(rddToCoPartitionWith.isSorted)
+      //the partition map to co-partition to
+      val collectedDestinationPartitionMap = rddToCoPartitionWith.partitionMap
+      //the partition map for <this>, used to infer the correct ReferenceRegion for every T
+      val collectedSourcePartitionMap = partitionMap
+      //number of partitions we will have after repartition
+      val partitions = collectedDestinationPartitionMap.length
+      // there are a lot of parts here, but in the end we get a repartitioned RDD where every
+      // T is copartitioned with the destination ReferenceRegions that it may join to
+      val finalPartitionedRDD = rdd.mapPartitionsWithIndex((idx, iter) => {
+        // First, we get the ReferenceRegion for every T
+        val listRepresentation = inferCorrectReferenceRegionsForPartition(iter, collectedSourcePartitionMap(idx)).toList
+        // Then we use this loop to key each T with the partition index(es) it belongs to
+        for (partition <- collectedDestinationPartitionMap.zipWithIndex) yield {
+          // The key becomes the destination partition
+          // The value is a list of (ReferenceRegion, T) that will end up in this partition
+          (partition._2, listRepresentation.filter(f => (f._1.start >= partition._1._1.start && f._1.start <= partition._1._2.end) ||
+            (f._1.end >= partition._1._1.start && f._1.end <= partition._1._2.end)))
+        }
+      }.toIterator.filter(_._2.nonEmpty), preservesPartitioning = true)
+        .partitionBy(new GenomicPositionRangePartitioner(partitions))
```

Comment:
Implementing this with `repartitionAndSortWithinPartitions` turning out to be a bit more difficult than I thought. We cannot define a partitioner that assigns a single key multiple partition indexes in spark. The code above is doing that (essentially).

As I am typing this, I think I've come up with a way to solve the problem. Hooray for answering one's own question.